rts/motoko-rts/src/region.rs

```diff
@@ -0,0 +1,593 @@
+        use crate::ic0_stable::nicer::{read_u64, write_u64};
+
+        fn index(r: &RegionId) -> u64 {
+            offset::REGION_TABLE + r.0 as u64 * size::REGION_TABLE_ENTRY as u64
+        }
+
+        /// Some(_) gives the size in pages.
+        /// None means that the region is not in use.
+        pub fn get(r: &RegionId) -> Option<RegionSizeInPages> {
+            RegionSizeInPages::from_u64(read_u64(index(r)))
+        }
+
+        pub fn set(r: &RegionId, s: Option<RegionSizeInPages>) {
+            write_u64(index(r), RegionSizeInPages::into_u64(s))
+        }
+    }
+}
+
+// Region manager's total memory size in stable memory, in _pages_.
+#[ic_mem_fn]
+pub unsafe fn region_mem_size<M: Memory>(_mem: &mut M) -> u64 {
+    if crate::memory::ic::REGION_MEM_SIZE_INIT {
+        meta_data::size::STATIC_MEM_IN_PAGES as u64
+            + crate::memory::ic::REGION_TOTAL_ALLOCATED_BLOCKS as u64
+                * (meta_data::size::PAGES_IN_BLOCK as u64)
+    } else {
+        // Before we initialization of anything, give back zero.
+        0
+    }
+}
+
+#[ic_mem_fn]
+pub unsafe fn region_new<M: Memory>(mem: &mut M) -> Value {
+    let r_ptr = mem.alloc_words(size_of::<Region>());
+
+    let next_id = meta_data::total_allocated_regions::get() as u16;
+    meta_data::total_allocated_regions::set(next_id as u64 + 1);
+
+    // NB. cannot use as_region() here as we didn't write the header yet
+    let region = r_ptr.get_ptr() as *mut Region;
+    (*region).header.tag = TAG_REGION;
+    (*region).id = next_id;
+    (*region).page_count = 0;
+    (*region).vec_pages = alloc_blob(mem, Bytes(0));
+
+    // Update Region table.
+    {
+        let r_id = RegionId::from_id((*region).id);
+        let c = meta_data::region_table::get(&r_id);
+        // sanity check: Region table says that this region is available.
+        assert_eq!(c, None);
+        meta_data::region_table::set(&r_id, Some(RegionSizeInPages(0)));
+    }
+    Value::from_ptr(region as usize)
+}
+
+#[ic_mem_fn]
+pub unsafe fn region_recover<M: Memory>(mem: &mut M, rid: &RegionId) -> Value {
+    use meta_data::size::PAGES_IN_BLOCK;
+    let c = meta_data::region_table::get(&rid);
+    let page_count = match c {
+        Some(RegionSizeInPages(pc)) => {
+            println!(
+                80,
+                "region_recover {:?} => found region record {:?}", rid, c
+            );
+            pc
+        }
+        _ => {
+            rts_trap_with("region_recover failed: Expected Some(RegionSizeInPages(_))");
+        }
+    };
+    let r_ptr = mem.alloc_words(size_of::<Region>());
+
+    // NB. cannot use as_region() here as we didn't write the header yet
+    let region = r_ptr.get_ptr() as *mut Region;
+    (*region).header.tag = TAG_REGION;
+    (*region).id = rid.0;
+    (*region).page_count = page_count as u32;
+
+    let block_count = (page_count as u32 + PAGES_IN_BLOCK - 1) / PAGES_IN_BLOCK;
+    (*region).vec_pages = alloc_blob(mem, Bytes(block_count));
```

Comment:
```suggestion
    (*region).vec_pages = alloc_blob(mem, Bytes(2*block_count));
```
Doesn't the blob need 2*block_count bytes to store block_count 2-byte u16 values?

Reply:
Indeed, thank you!

Reply:
It may not be the bug I have now, since the `block_count` is supposed to be zero for region 0.  I'll fix this and see what happens though.

Reply:
Though this was a lurking bug for later, like I thought, it didn't address the bug for `stable-mem-nat8`.

I added a couple of more debug print statements into the RTS and now am getting convinced that the GC is moving my heap-allocated region0 object that the RTS puts into a global variable.

I can have a variable in the program that points at it, and by printing out the pointer as a `usize` and looking at the value, I see that they start by agreeing, but then disagree when the variable gets re-directed but the RTS-held version is not re-directed.  (I say "redirected" because its value changes after the initialization is over, as the pre-upgrade is happening).

WDYT?

Reply:
Note for readers -- Figured this out with @crusso's help offline.